src/libraries/System.Private.CoreLib/src/System/IO/Strategies/FileStreamHelpers.Windows.cs

```diff
@@ -225,40 +225,41 @@ internal static async Task AsyncModeCopyToAsync(SafeFileHandle handle, bool canS
                         }
 
                         // If the operation did not synchronously succeed, it either failed or initiated the asynchronous operation.
-                        if (!synchronousSuccess)
+                        if (!synchronousSuccess && errorCode != Interop.Errors.ERROR_IO_PENDING)
                         {
-                            switch (errorCode)
+                            if (RandomAccess.IsEndOfFile(errorCode, handle, readAwaitable._position))
                             {
-                                case Interop.Errors.ERROR_IO_PENDING:
-                                    // Async operation in progress.
-                                    break;
-                                case Interop.Errors.ERROR_BROKEN_PIPE:
-                                case Interop.Errors.ERROR_HANDLE_EOF:
-                                    // We're at or past the end of the file, and the overlapped callback
-                                    // won't be raised in these cases. Mark it as completed so that the await
-                                    // below will see it as such.
-                                    readAwaitable.MarkCompleted();
-                                    break;
-                                default:
-                                    // Everything else is an error (and there won't be a callback).
-                                    throw Win32Marshal.GetExceptionForWin32Error(errorCode, handle.Path);
+                                // We're at or past the end of the file, and the overlapped callback
+                                // won't be raised in these cases. Mark it as completed so that the await
+                                // below will see it as such.
+                                readAwaitable.MarkCompleted();
+                            }
+                            else
+                            {
+                                // Everything else is an error (and there won't be a callback).
+                                throw Win32Marshal.GetExceptionForWin32Error(errorCode, handle.Path);
                             }
```

Comment:
Nit: I'd personally find this cleaner if the if were inverted:
```C#
if (!RandomAccess.IsEndOfFile(errorCode, handle, readAwaitable._position))
{
    throw Win32Marshal.GetExceptionForWin32Error(errorCode, handle.Path);
}

// We're at or past the end of the file, and the overlapped callback
// won't be raised in these cases. Mark it as completed so that the await
// below will see it as such.
readAwaitable.MarkCompleted();
```